### PR TITLE
8341975: Unable to set encoding for IO.println, IO.print and IO.readln

### PR DESCRIPTION
**Pull request by naotoj**
Changing the charset initialization of `java.io.Console` class, which is the basis of `java.io.IO`, so that it would behave the same way as `System.out` wrt encoding. This change will also have the capability to override the default charset used in `IO` methods with `stdout.encoding` system property. A corresponding CSR has also been drafted.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8342208](https://bugs.openjdk.org/browse/JDK-8342208) to be approved

### Issues
 * [JDK-8341975](https://bugs.openjdk.org/browse/JDK-8341975): Unable to set encoding for IO.println, IO.print and IO.readln (**Enhancement** - P4)
 * [JDK-8342208](https://bugs.openjdk.org/browse/JDK-8342208): Unable to set encoding for IO.println, IO.print and IO.readln (**CSR**)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**) Review applies to [805365c6](https://git.openjdk.org/jdk/pull/21569/files/805365c6e32270b049cc1c34da8497096b89ed9b)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) Review applies to [805365c6](https://git.openjdk.org/jdk/pull/21569/files/805365c6e32270b049cc1c34da8497096b89ed9b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21569/head:pull/21569` \
`$ git checkout pull/21569`

Update a local copy of the PR: \
`$ git checkout pull/21569` \
`$ git pull https://git.openjdk.org/jdk.git pull/21569/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21569`

View PR using the GUI difftool: \
`$ git pr show -t 21569`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21569.diff">https://git.openjdk.org/jdk/pull/21569.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21569#issuecomment-2420449027)

### CODE REVIEW

**Review by RogerRiggs — APPROVED** [review]
Looks good.

**Review by irisclark — APPROVED** [review]

File: src/java.base/share/classes/java/io/Console.java [hunk 1]
```diff
@@ -580,7 +580,8 @@ public void flush() {
      * the {@code Console}.
      * <p>
      * The returned charset corresponds to the input and output source
-     * (e.g., keyboard and/or display) specified by the host environment or user.
+     * (e.g., keyboard and/or display) specified by the host environment or user,
+     * which defaults to the one based on {@link System##stdout.encoding stdtout.encoding}.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/Console.java:584]: In System.getProperties, this is specified for as the encoding for System.out. That may need to be expanded to say Console too.

Comment (naotoj) [reply src/java.base/share/classes/java/io/Console.java:584]: Added `System.console()` mention to `stdout.encoding`. Also removed `Console` mention from `System.out`, as it no longer provides extra meaning.

**Review by irisclark — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]

**Review by RogerRiggs — APPROVED** [review]

File: src/java.base/share/classes/java/lang/System.java [hunk 2]
```diff
@@ -147,8 +147,7 @@ private System() {
      * corresponds to display output or another output destination
      * specified by the host environment or user. The encoding used
      * in the conversion from characters to bytes is equivalent to
-     * {@link Console#charset()} if the {@code Console} exists,
-     * <a href="#stdout.encoding">stdout.encoding</a> otherwise.
+     * <a href="#stdout.encoding">stdout.encoding</a>.
```
Comment (liach) [inline src/java.base/share/classes/java/lang/System.java:150]:
```suggestion
     * {@systemProperty stdout.encoding}.
```

This defines the anchor with the same id, and also makes this property available for javadoc index search. Example include the system property at https://download.java.net/java/early_access/jdk24/docs/api/java.base/java/lang/ClassLoader.html#java.system.class.loader

Comment (jaikiran) [reply src/java.base/share/classes/java/lang/System.java:150]: Hello Chen, the `{@systemProperty}` tag is expected to be used where the system property is being defined (the "Where should the tag be used?" section in https://mail.openjdk.org/pipermail/core-libs-dev/2018-November/056653.html). 
On this line, it's merely being referenced and the actual definition of this property is in the documentation of System.getProperties() method and there it correctly already uses a `{@systemProperty}` tag for this property https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/System.java#L790

Comment (liach) [reply src/java.base/share/classes/java/lang/System.java]: Indeed my bad, this is an href instead of an id.

**Review by RogerRiggs — APPROVED** [review]
lgtm

**Comment by naotoj** [comment]
Thank you for your reviews!
/integrate

**Pull request closed without merging**